02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,279 @@
+
+## Motivation
+
+We're going to start today with some Ruby code that does something a little unexpected. It's a method that takes an array of strings as an argument, and truncates (chops off the end of) all the strings with more than three characters. Or at least that's the idea.
+
+```ruby
+def short_strings(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings(pets)
+puts "#{pets}"
+
+```
+
+Running this code results in `["dog", "parrot", "cat", "llama"]`. The array in unchanged! Let's do some debugging:
+
+```ruby
+def short_strings(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+  puts "Inside short_strings, input is"
+  puts "#{input}"
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings(pets)
+puts "After calling short_strings"
+puts "#{pets}"
+```
+
+The results:
+
+```
+Inside short_strings, input is
+["dog", "par", "cat", "lla"]
+After calling short_strings
+["dog", "parrot", "cat", "llama"]
+```
+
+`short_strings` is indeed creating a list of shortened words and storing it under the name `input`, but our outer variable `pets` isn't being updated. The reason why has to do with references and values, and how data is stored in a computer.
+
+As an aside: one way to fix our method is to return the new array, and when calling it say `pets = short_strings(pets)`. However, sometimes this isn't an option - for example, what if our method was supposed to return the number of words that had to be truncated?
+
+## References and Values
+
+When we create an array in Ruby (or a string or a hash or any other complex data type), we're actually creating two things.
+
+The first is the _value_ of the array, which involves asking the operating system for a bit of memory and then putting our data in it. You can think of this as the actual object. Each piece of memory we get from the OS has an _address_ representing its physical location in hardware, which is how we get back to it later.
+
+The second is a _reference_ to the array, which ties together the address of that memory with a name for our program to use. References are sometimes called _pointers_ (especially in C), and we say that a variable _points to_ or _references_ an object.
+
+Recalling our example above, saying
+
+```ruby
+pets = ['dog', 'parrot', 'cat', 'llama']
+```
+
+and imagining that the `pets` array has been stored at address `1234`, we would get the following memory layout:
+
+![references and variables](images/references_and_values/references_and_values.png)
+
+Every variable in Ruby consists of these two parts, a reference and a value. Normally when you type the variable's name, Ruby automatically goes and gets the object. If you want to find out the address, you can use the `object_id` method:
```

Comment:
I realize I'm being a bit pedantic - so feel free to push back. :) I'm struggling with saying that _object_id_ returns the _address_. I understand that for most objects, object_id in Ruby would be the same as VALUE that represents the object at the C level. But there are special cases for true, false, nil and FIXNUMs in Ruby (very clever special case, but special case none the less). 
Could we change this to something like; "If you want to find out which object you are working with, you can use the 'object_id' method:"